tools/update/docs/lepton-update.1.in
.TH lepton-update 1 "@DATE@" "Lepton EDA" @VERSION@
.SH NAME
lepton-update - update Lepton EDA symbols and schematics
.SH SYNOPSIS
.B lepton-update [\fIOPTION\fR ...] \fIFILE\fR ...

.SH OPTIONS
.TP 8
\fB-h\fR, \fB--help\fR
Print a help message.
.TP 8
\fB-V\fR, \fB--version\fR
Print version information.

.SH DESCRIPTION
.PP

\fBlepton-update\fR is a tool for updating old schematic and
symbols.

.PP
The program takes a schematic or symbol filenames on the command
line and updates them after making backups.  It is intended for
running against schematics which are either version 20030921 or
earlier.

.PP
For both symbols and schematics, the program fixes obsolete text
and pin format lines and sets the current format version.

.PP
For schematics, it replaces the \fIlabel=\fR attribute on nets with
\fInetname=\fR, and \fIuref=\fR with \fIrefdes=\fR.

For symbols, it changes them as follows:

.IP \[bu] 2
Converts all \fIpin#=#\fR attributes into \fIpinseq=\fR and
\fIpinnumber=\fR attributes.
.IP \[bu]
Converts all \fIslot#=#\fR attributes into \fIslotdef=\fR
attributes.
.IP \[bu]
Replaces pin attributes \fItype=\fR with \fIpintype=\fR.
.IP \[bu]
Replaces pin attributes \fIlabel=\fR with \fIpinlabel=\fR.
.IP \[bu]
Replaces symbol attribute \fIuref=\fR with \fIrefdes=\fR.

.SH AUTHORS
Vladimir Zhbanov

.SH COPYRIGHT
.nf
Copyright \(co 2021-@YEAR@ Lepton Developers.
License GPLv2+: GNU GPL version 2 or later. Please see the `COPYING'
file included with this program for full details.
.PP
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

.SH SEE ALSO
\fBlepton-schematic\fR(1), \fBlepton-symcheck\fR(1)


NEWS
==============================================
 'gschem and Friends' Electronic Design Suite
==============================================

Copyright (C) 1998-2016 gEDA Developers

This file documents important user-visible changes in gEDA/gaf.  For
more information, please consult the `ChangeLog' file.

Notable changes in gEDA/gaf 1.9.3
====================================

* The gEDA/gaf source tarball no longer includes libintl.  When
  building gEDA/gaf with internationalisation support, a GNU
  gettext-compatible libintl is now required.

* gEDA/gaf core tools will now automatically rotate per-user log
  files.  250 log files per tool will be retained from the last 24
  hours, and 50 log files per tool from all time.

* gEDA/gaf tools now search for data and configuration files in the
  standard XDG directories.  By default, these are:

  - `/usr/local/share/gEDA/` and `/usr/share/gEDA` for system data
  - `/etc/xdg/gEDA/` for system configuration
  - `$HOME/.local/share/gEDA/` for per-user data
  - `$HOME/.config/gEDA/` for per-user configuration.

  However, gEDA/gaf tools will still use `$HOME/.gEDA` for per-user
  files if that directory exists, they still obey the `$GEDADATA` and
  `$GEDADATARC` environment variables, and by default the search path
  includes the installation prefix selected when compiling gEDA/gaf.

* The configure script now supports an `--enable-relocatable` switch.
  When specified, gEDA/gaf tools do not have the configured prefix
  compiled into them.

* The configure script now supports a `--disable-deprecated` switch.
  When specified, various deprecated features and behaviour are
  disabled, which may break backward-compatibility with some
  configurations.

* A new Scheme API function, `log!`, has been added to the `(geda
  log)` module.  It allows Scheme code to emit log messages in the
  same way that the tools' C code does.

* Gschem multiattrib dialog improvements:

  - Shift+Tab can be used to move focus backward from `value' to
    `name'.

  - Tooltip to the `value' text entry has been added explaning how
    to enter newline or tab.

  - Horizontal scrollbar for `value' has been added to prevent
    resizing of the dialog.

* Gschem undo information now contains center of viewport and
  current page scale, so if gschem window size changes before
  undo, page will be zoomed accordingly.

* Faster redrawing of pages in gschem.

* Menu entry to launch Scheme REPL in gschem's background terminal
  has been added.

* Changed calculation of the size of bus and net junctions.

* `Bom` gnetlist backend has been refactored, all its functions
  have help now.

* View bounds in gschem window has been changed so symbols can be
  edited without translation.

* New Scheme module for part list creation has been added for
  gnetlist.

* Netlist output of several gnetlist backends is sorted by package
  names.

* New Scheme procedures for working with source library have been
  added to the `(geda library)` module: source-library-contents
  and set-source-library-contents!.  Please see the "geda-scheme"
  info manual for more details.


Notable changes in gEDA/gaf 1.9.2
=================================
* Build system changes:

  - Guile 2.0 or newer is now required.

  - top-level control file allowing continuous integration using
    Travis CI has been added.

* Nullor expansion has been fixed in the `spice' and `spice-sdb'
  backends.

  - Element name prefix "E_" has been substituted for "E-" to prevent
    errors output by some simulators (e.g. ngspice).

  - A default gain value has been added in the nullor expansion code.

* Two obsolete scripts verilog2vhdl.sh and script.sed have been
  removed.

* Undocumented and ambiguous feature of the `spice-sdb' backend for
  gnetlist consisting in the possibility of using the "value="
  attribute instead of "file=" only for include blocks and only in
  embedding mode has been removed.

* The following scripts and programs have been moved from the `utils'
  directory to a new `contrib' directory and are no longer installed
  by default:

         smash_megafile          olib
         gmk_sym                 convert_sym
         sarlacc_schem           sarlacc_sym
         pads_backannotate       sw2asc
         gnet_hier_verilog.sh

* `gpstoimage' has been removed from the distribution; equivalent
  functionality is provided by the `gaf' command.

* `convert_sym.awk' has been removed from the distribution; equivalent
  functionality is provided by the `convert_sym' command.

* gnetlist now supports output of netlists to the standard output.

* New Scheme procedures and hooks have been added.  Please see the
  "geda-scheme" info manual for more details.

  - `set-component-with-transform!' is a modified version of
    `set-component!', which applies all the specified transformations
    to a component and immediately updates it.

  - `auto-refdes-reset!' resets components' refdeses

  - `copy-objects-hook' is called whenever an object is copied, it can
    be used, for example, to reset component's refdes on copy.

* The following gschem configuration items have been added:

  - `third-button-cancel' cancels draw actions by the third mouse
    button in mousepan mode.

  - `grid-mode' sets default grid mode.

* Fixes/updates to the *BSD, Mac OS X, and MinGW builds.

* Changes in the gschem UI event state machine:

  - New drawing modes apply immediately.

  - Indication of an action being in progress has been added.

  - Mirror action can now be used during an object placement.

  - Show/hide text function is now enabled during other actions.

  - RMB panning is now enabled during placing or drawing.

* Some gschem dialog have been merged, made non-modal, and placed in
  sidebar:

  - Color, line, fill and pin type dialogs have been merged into one
    non-modal object properties dialog with widgets applying values
    immediately.

  - Text properties dialog.


utils/update/lepton-update.scm
;;; lepton-update - Update schematics and symbols in the gEDA/gaf
;;; file format.
;;;
;;; Copyright (C) 2021-2022 Lepton EDA Contributors.
;;;
;;; This program is free software; you can redistribute it and/or modify
;;; it under the terms of the GNU General Public License as published by
;;; the Free Software Foundation; either version 2 of the License, or
;;; (at your option) any later version.
;;;
;;; This program is distributed in the hope that it will be useful,
;;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;;; GNU General Public License for more details.
;;;
;;; You should have received a copy of the GNU General Public License
;;; along with this program; if not, write to the Free Software
;;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

(use-modules (ice-9 match)
             (ice-9 rdelim)
             (ice-9 regex)
             (srfi srfi-11)
             (lepton attrib)
             (lepton ffi)
             (lepton file-system)
             (lepton gettext)
             (lepton object)
             (lepton page)
             (lepton srfi-37)
             (lepton toplevel)
             (lepton version)
             (symbol check obsolete))

;;; Initialize liblepton library.
(liblepton_init)
(unless (getenv "LEPTON_INHIBIT_RC_FILES")
  (register-data-dirs))


(define (usage)
  (define program-name (basename (car (program-arguments))))

  (format #t "Usage: ~A [OPTION ...] FILE ...

Update files in gEDA/gaf format up to the current version.

Command line switches:
  -h,--help             -- Print usage information.
  -V,--version          -- Print version information.
"
          program-name)
  (exit 0))


;;; Define uninterned symbol to aid with setting of boolean
;;; 'visible?' value.
(define *visible?* (make-symbol "*visible?*"))

(define* (copy-attrib! object
                       #:key
                       (anchor #f)
                       (align #f)
                       (angle #f)
                       (string #f)
                       (size #f)
                       (visible? *visible?*)
                       (show #f))
  (let ((new-object (copy-object object)))
    (set-text! object
               (or anchor (text-anchor object))
               (or align (text-align object))
               (or angle (text-angle object))
               (or string (text-string object))
               (or size (text-size object))
               (if (eq? visible? *visible?*)
                   (text-visible? object)
                   visible?)
               (or show (text-attribute-mode object)))))

(define (set-attrib-name! attrib name)
  (set-text-string! attrib
                    (string-append name "=" (attrib-value attrib))))

(define (attrib-name= object name)
  (and (attribute? object)
       (string= (attrib-name object) name)))

(define (fix-symbol-attrib object)
  (define (fix-slot#-attrib! object)
    ;; Fix slot#=# attribute.
    (let ((num (string-drop (attrib-name object) 4)))
      (set-attrib-name! object "slotdef")
      (set-attrib-value! object
                         (string-append num ":" (attrib-value object)))))

  (define (fix-pin#-attrib! object)
    ;; Numbers in name and value of pin#=# attributes are
    ;; always the same, e.g. pin10=10.  So we just change
    ;; attribute name here.
    (set-attrib-name! object "pinnumber")
    ;; Make 'pinseq' attribute.
    (let ((pinseq (copy-attrib! object #:show 'both #:visible? #f))
          (pin (attrib-attachment object)))
      (set-attrib-name! pinseq "pinseq")
      (attach-attribs! pin pinseq)))

  (cond
   ((obsolete-slot#-attrib? object) (fix-slot#-attrib! object))
   ((obsolete-pin#-attrib? object) (fix-pin#-attrib! object))
   ((attrib-name= object "uref") (set-attrib-name! object "refdes"))
   ((attrib-name= object "type") (set-attrib-name! object "pintype"))
   ((attrib-name= object "label") (set-attrib-name! object "pinlabel"))
   (else #f)))

(define (fix-schematic-attrib object)
  (cond
   ((attrib-name= object "label") (set-attrib-name! object "netname"))
   ((attrib-name= object "uref") (set-attrib-name! object "refdes"))
   (else #f)))


(define (files-ok? filename backup)
  (let ((msg
         (cond
          ((or (not (file-exists? filename))
               (not (file-readable? filename)))
           (format #f "File ~A does not exist or is not readable.  Skipping.\n" filename))
          ((directory? filename)
           (format #f "~A is a directory, not a file.  Skipping.\n" filename))
          ((file-exists? backup)
           (format #f "Found backup file: ~A.  Skipping ~A\n" backup filename))
          (else #f))))
    (when msg
      (display msg (current-error-port)))
    (not msg)))

(define (filename->fix-func filename)
  (cond
   ((string-suffix-ci? ".sch" filename)
    fix-schematic-attrib)
   ((string-suffix-ci? ".sym" filename)
    fix-symbol-attrib)
   (else (format (current-error-port) "Could not determine the type of file ~S\n" filename)
         (exit 1))))

(define (page->file page filename)
  (with-output-to-file filename
    (lambda () (display (page->string page)))))

(define (update filename)
  (let ((backup (string-append filename ".bak"))
        (fix-func (filename->fix-func filename)))
    (when (files-ok? filename backup)
      (rename-file filename backup)
      (format (current-error-port) "Updating: ~A (backup: ~A)\n" filename backup)

      (let ((page (file->page backup)))
        (for-each fix-func (page-contents page))
        (page->file page filename)))))


;;; Main program.
(when (= (length (program-arguments)) 1) (usage))

;;; Parse lepton-schematic command-line options, displaying usage
;;; message or version information as required.
(define (parse-commandline)
  "Parse command line options.  Return the list of non-option
arguments which should represent the list of schematic and symbol
files to process."
  (reverse
   (args-fold
    (cdr (program-arguments))
    (list
     (option '(#\h #\? "help") #f #f
             (lambda (opt name arg seeds)
               (usage)))
     (option '(#\V "version") #f #f
             (lambda (opt name arg seeds)
               (display-lepton-version #:print-name #t #:copyright #t)
               (exit 0))))
    (lambda (opt name arg seeds)
      (format #t
              (G_ "ERROR: Unknown option ~A.
Run `~A --help' for more information.\n")
              (if (char? name)
                  (string-append "-" (char-set->string (char-set name)))
                  (string-append "--" name))
              (basename (car (program-arguments))))
      (exit 1))
    (lambda (op seeds) (cons op seeds))
    '())))


(with-toplevel
 (make-toplevel)
 (lambda ()
   (let ((files (parse-commandline)))
     (for-each update files))))
